### ISSUE TITLE
paddle2.3.1.post112

### ISSUE THREAD

**Issue by HUSTHY** [PaddlePaddle/Paddle#44981]
### bug描述 Describe the Bug

‘’‘
python
from paddlenlp.transformers import BertPretrainedModel,BertModel
import paddle.nn as nn
import paddle

class SentenceBert(BertPretrainedModel):
    """
    __init__(self,bert) 一定要带一个参数，外层from_pretrained()才能成功
    bert会在from_pretrained()的时候初始化成功然后传入
    """
    # base_model_prefix = "bert"
    def __init__(self, bert):
        super(SentenceBert,self).__init__()
        self.bert = bert
        self.classifier = nn.Linear(3*768,2)

    def forward(self, inputs, **kwargs):
        input_a = inputs[0]
        input_b = inputs[1]
        output_a = self.bert(**input_a, output_hidden_states=True)
        output_b = self.bert(**input_b, output_hidden_states=True)
        # 采用最后一层
        embedding_a = output_a[0][-1]
        embedding_b = output_b[0][-1]
        embedding_a = self.pooling(embedding_a, input_a)
        embedding_b = self.pooling(embedding_b, input_b)

        embedding_abs = paddle.abs(embedding_a - embedding_b)
        vectors_concat = []
        vectors_concat.append(embedding_a)
        vectors_concat.append(embedding_b)
        vectors_concat.append(embedding_abs)
        # 列拼接3个768————>3*768
        features = paddle.concat(vectors_concat,axis=-1)
        output = self.classifier(features)
        return output

    def pooling(self, token_embeddings, input):
        output_vectors = []
        # attention_mask
        attention_mask = input['attention_mask']
        # [B,L]------>[B,L,1]------>[B,L,768],矩阵的值是0或者1
        # input_mask_expanded = attention_mask.unsqueeze(-1).expand(token_embeddings.size()).float()
        input_mask_expanded = paddle.unsqueeze(attention_mask,-1)
        input_mask_expanded = paddle.expand(input_mask_expanded,shape=[input_mask_expanded.shape[0],input_mask_expanded.shape[1],token_embeddings.shape[-1]])

        # 这里做矩阵点积，就是对元素相乘(序列中padding字符，通过乘以0给去掉了)[B,L,768]
        t = token_embeddings * input_mask_expanded

        # [B,768]
        sum_embeddings = paddle.sum(t, 1)

        # [B,768],最大值为seq_len
        sum_mask = input_mask_expanded.sum(1)
        # 限定每个元素的最小值是1e-9，保证分母不为0
        sum_mask = paddle.clip(sum_mask, min=1e-9)
        # 得到最后的具体embedding的每一个维度的值——元素相除
        output_vectors.append(sum_embeddings / sum_mask)

        # 列拼接
        output_vector = paddle.concat(output_vectors, 1)

        return output_vector

import paddle
from data_reader.dataReader import DataReader
from model.sentence_bert import SentenceBert
import argparse
from tools.progressbar import ProgressBar
from tools.log import Logger
from datetime import datetime
import os
from paddlenlp.transformers import BertTokenizer
from paddle.io import DataLoader
from paddle.optimizer import AdamW
from paddle.optimizer.lr import ReduceOnPlateau

import paddle.nn.functional  as F
import numpy as np
import random
from tqdm import tqdm

# os.environ['FLAGS_cudnn_deterministic'] = "1"

paddle.seed(100)
random.seed(100)
np.random.seed(100)

logger = Logger('paddle_sbert_loger',log_level=10,log_file='./log_output/pawsx_sbert_paddle.log').logger
paddle.set_device('gpu:1')

def parse_args():
    parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    parser.add_argument("--max_len",type=int,default=64)
    parser.add_argument("--train_file", type=str,default='./data/paws_x/translated_train.tsv', help="train text file")
    parser.add_argument("--val_file", type=str, default='./data/paws_x/dev_2k.tsv',help="val text file")
    parser.add_argument("--pretrained", type=str, default="pretrained_models/paddle/bert-wwm-ext-chinese", help="huggingface pretrained model")
    parser.add_argument("--model_out", type=str, default="./output", help="model output path")
    parser.add_argument("--batch_size", type=int, default=64, help="batch size")
    parser.add_argument("--epochs", type=int, default=1, help="epochs")
    parser.add_argument("--lr", type=int, default=1e-5, help="epochs")
    parser.add_argument("--task_type",type=str,default='classification')
    args = parser.parse_args()
    return args

def collate_fn(batch):
    input_ids_a, attention_mask_a, token_type_id_a, input_ids_b, attention_mask_b, token_type_id_b, label = zip(*batch)
    bz = len(input_ids_a)
    max_a = 0
    for i in range(len(input_ids_a)):
        if input_ids_a[i].shape[0] > max_a:
            max_a = input_ids_a[i].shape[0]

    input_ids_a_pad = paddle.zeros(shape=[bz,max_a],dtype="int64")
    for i in range(len(input_ids_a)):
        l = input_ids_a[i].shape[0]
        input_ids_a_pad[i,0:l] = input_ids_a[i]

    attention_mask_a_pad = paddle.zeros(shape=[bz, max_a],dtype="int64")
    for i in range(len(input_ids_a)):
        l = attention_mask_a[i].shape[0]
        attention_mask_a_pad[i, 0:l] = attention_mask_a[i]

    token_type_id_a_pad = paddle.zeros(shape=[bz, max_a],dtype="int64")
    for i in range(len(input_ids_a)):
        l = token_type_id_a[i].shape[0]
        token_type_id_a_pad[i, 0:l] = token_type_id_a[i]

    max_b = 0
    for i in range(len(input_ids_b)):
        if input_ids_b[i].shape[0] > max_b:
            max_b = input_ids_b[i].shape[0]

    input_ids_b_pad = paddle.zeros(shape=[bz, max_b],dtype="int64")
    for i in range(len(input_ids_b)):
        l = input_ids_b[i].shape[0]
        input_ids_b_pad[i, 0:l] = input_ids_b[i]

    attention_mask_b_pad = paddle.zeros(shape=[bz, max_b],dtype="int64")
    for i in range(len(input_ids_b)):
        l = attention_mask_b[i].shape[0]
        attention_mask_b_pad[i, 0:l] = attention_mask_b[i]

    token_type_id_b_pad = paddle.zeros(shape=[bz, max_b],dtype="int64")
    for i in range(len(input_ids_b)):
        l = token_type_id_b[i].shape[0]
        token_type_id_b_pad[i, 0:l] = token_type_id_b[i]

    label = paddle.stack(label,axis=0)
    return input_ids_a_pad, attention_mask_a_pad, token_type_id_a_pad, input_ids_b_pad, attention_mask_b_pad, token_type_id_b_pad, label

def train(args):
    logger.info("args: %s",args)
    #会自动下载相应的权重和配置文件
    tokenizer = BertTokenizer.from_pretrained('bert-wwm-ext-chinese')
    model = SentenceBert.from_pretrained('bert-wwm-ext-chinese')

    # tokenizer = BertTokenizer.from_pretrained(pretrained_model_name_or_path=args.pretrained)
    # model = SentenceBert.from_pretrained(pretrained_model_name_or_path=args.pretrained)

    train_dataset = DataReader(tokenizer=tokenizer,filepath=args.train_file,max_len=args.max_len)
    train_dataloader = DataLoader(train_dataset,batch_size=args.batch_size,shuffle=False, collate_fn=collate_fn)

    val_dataset = DataReader(tokenizer=tokenizer,filepath=args.val_file,max_len=args.max_len)
    val_dataloader = DataLoader(val_dataset, batch_size=args.batch_size, shuffle=False, collate_fn=collate_fn)

    optimizer = AdamW(parameters = model.parameters(),learning_rate = args.lr)
    scheduler = ReduceOnPlateau(learning_rate= args.lr,mode='max',factor=0.5, patience=5)

    logger.info("***** Running training *****")
    logger.info("  Num examples = %d", len(train_dataloader))
    logger.info("  Num Epochs = %d", args.epochs)

    time_srt = datetime.now().strftime('%Y-%m-%d')
    save_path = os.path.join(args.model_out,"paddle_" + time_srt)

    best_acc = 0.0
    patience = 5
    no_best_epoch = 0
    model.train()
    for epoch in range(args.epochs):
        pbar = ProgressBar(n_total=len(train_dataloader), desc='Training')
        model.train()
        for step,batch in enumerate(train_dataloader):
            inputs_a = {'input_ids':batch[0],'attention_mask':batch[1],'token_type_ids':batch[2]}
            inputs_b = {'input_ids': batch[3], 'attention_mask': batch[4], 'token_type_ids': batch[5]}
            labels = batch[6]
            inputs = []
            inputs.append(inputs_a)
            inputs.append(inputs_b)
            output = model(inputs)
            loss = F.cross_entropy(output,labels)
            loss.backward()
            optimizer.step()
            optimizer.clear_grad()
            pbar(step, {'loss':loss.item()})

        val_acc = valdation(model, val_dataloader)
        scheduler.step(val_acc)

        if val_acc > best_acc:
            best_acc = val_acc

            if not os.path.exists(save_path):
                os.makedirs(save_path)
            logger.info("save model")
            model.save_pretrained(save_path)
            tokenizer.save_pretrained(save_path)
            no_best_epoch = 0
        else:
            no_best_epoch += 1
        logger.info("val_acc:%.4f------best_acc:%.4f" % ( val_acc, best_acc))

        if no_best_epoch >= patience:
            logger.info("taining finished because of no improving" )
            exit()

    
def valdation(model,val_dataloader):
    total = 0
    total_correct = 0
    model.eval()
    with paddle.no_grad():
        pbar = ProgressBar(n_total=len(val_dataloader), desc='evaldation')
        for step, batch in enumerate(val_dataloader):
            inputs_a = {'input_ids': batch[0], 'attention_mask': batch[1], 'token_type_ids': batch[2]}
            inputs_b = {'input_ids': batch[3], 'attention_mask': batch[4], 'token_type_ids': batch[5]}
            labels = paddle.squeeze(batch[6])
            inputs = []
            inputs.append(inputs_a)
            inputs.append(inputs_b)
            output = model(inputs)
            pred = paddle.argmax(output, axis=1)
            correct = (labels == pred).sum()
            total_correct += correct
            total += labels.shape[0]
            loss = F.cross_entropy(output, labels)
            pbar(step, {'loss': loss.item()})

    acc = total_correct / total
    return acc

def main():
    args =parse_args()
    train(args)

if __name__ == '__main__':
    main()

’‘’

### 其他补充信息 Additional Supplementary Information

使用paddle2.3.1.post112/cuda11.2/3090显卡/paddlenlp2.3.4/Linux/python3.7环境复现SBert的时候，固定随机种子
paddle.seed(100)
random.seed(100)
np.random.seed(100)
并且使用FLAGS_cudnn_deterministic = True结果仍然不能复现，每次训练的acc指标都不能loss也不同；
同torch版本对比，在同一个数据集上(paws_x)，同一个Bert预训练权重 bert-wwm-ext-chinese下：
torch则可以复现，并且torch的acc为0.75(torch不设置随机种子的时候结果也是瞎飞)，paddle的acc0.55-0.62之间随机出现；
请问是环境版本问题还是paddle框架问题还是我代码问题？怎么解决？

**Comment by JZ-LIANG** [comment]
loss 复现问题请先排查一下四个方向：
0. 关闭模型中所有可能产生的随机的layer，看是否能复习：比如关闭 dropout， random sample 等
1. 模型参数初始化是否相同： 可以通过在训练第一个step 前打印所有参数查看
2. 模型更新设置是否相同：比如 lr 设置，优化器设置，grad clip 设置
3. 训练数据是否固定： 比较前后两次训练中，dataloader 出来的 ids，label 等数据是否一致

**Comment by HUSTHY** [comment]
排查过了的
0、我的组网代码里没有能产生随机的layer，只有一个bert和一个全连接层
1、没有排查不过应该也是差不太多，前几个epoch是一样的loss是一样的，后面就不一样了
2、从代码来看都是同一个设置，lr初始值一样、优化器一样、没有梯度裁剪
3、是一样的
CPU模式下运行结果可复现

**Comment by JZ-LIANG** [comment]
0.粒度太大了，bert 中是有 dropout 的
4. 如果上述几个点确定没有问题，只能“逐层核”对了。 具体是逐层打印每一层的 前向& 反向的 output，找到第一个对不齐的layer，反馈给我们，我们针对这个layer 来排查。

**Comment by HUSTHY** [comment]
已收到！谢谢！                          ——黄洋
